File: packages/git-storage-sdk-node/tests/webhook.test.ts
import { describe, expect, it } from 'vitest';

import {
  type RawWebhookPushEvent,
  type WebhookPushEvent,
  parseSignatureHeader,
  validateWebhook,
  validateWebhookSignature,
} from '../src';
import { createHmac } from '../src/util';

describe('Webhook Validation', () => {
  const secret = 'test_webhook_secret_key_123';
  const rawPayload: RawWebhookPushEvent = {
    repository: {
      id: 'repo_abc123def456ghi789jkl',
      url: 'https://git.example.com/org/repo',
    },
    ref: 'main',
    before: 'abc123000000000000000000000000000000000',
    after: 'def456000000000000000000000000000000000',
    customer_id: 'cust_xyz789mno456pqr123st',
    pushed_at: '2024-01-20T10:30:00Z',
  };
  const expectedPushPayload: WebhookPushEvent = {
    type: 'push',
    repository: {
      id: 'repo_abc123def456ghi789jkl',
      url: 'https://git.example.com/org/repo',
    },
    ref: 'main',
    before: 'abc123000000000000000000000000000000000',
    after: 'def456000000000000000000000000000000000',
    customerId: 'cust_xyz789mno456pqr123st',
    pushedAt: new Date('2024-01-20T10:30:00Z'),
    rawPushedAt: '2024-01-20T10:30:00Z',
  };
  const payloadStr = JSON.stringify(rawPayload);

  // Helper to generate a valid signature
  async function generateSignature(
    payloadData: string,
    webhookSecret: string,
    timestamp?: number
  ): Promise<{ header: string; timestamp: number }> {
    const ts = timestamp ?? Math.floor(Date.now() / 1000);
    const signedData = `${ts}.${payloadData}`;
    const signature = await createHmac('sha256', webhookSecret, signedData);
    return {
      header: `t=${ts},sha256=${signature}`,
      timestamp: ts,
    };
  }

  describe('parseSignatureHeader', () => {
    it('should parse valid signature header', () => {
      const header = 't=1234567890,sha256=abcdef123456';
      const result = parseSignatureHeader(header);
      expect(result).toEqual({
        timestamp: '1234567890',
        signature: 'abcdef123456',
      });
    });

    it('should handle header with spaces', () => {
      const header = 't=1234567890, sha256=abcdef123456';
      const result = parseSignatureHeader(header);
      expect(result).toEqual({
        timestamp: '1234567890',
        signature: 'abcdef123456',
      });
    });

    it('should return null for invalid header format', () => {
      expect(parseSignatureHeader('')).toBeNull();
      expect(parseSignatureHeader('invalid')).toBeNull();
      expect(parseSignatureHeader('t=123')).toBeNull(); // Missing signature
      expect(parseSignatureHeader('sha256=abc')).toBeNull(); // Missing timestamp
      expect(parseSignatureHeader('timestamp=123,signature=abc')).toBeNull(); // Wrong keys
    });

    it('should handle header with extra fields', () => {
      const header = 't=1234567890,sha256=abcdef123456,v1=ignored';
      const result = parseSignatureHeader(header);
      expect(result).toEqual({
        timestamp: '1234567890',
        signature: 'abcdef123456',
      });
    });
  });

  describe('validateWebhookSignature', () => {
    it('should validate correct signature', async () => {
      const { header, timestamp } = await generateSignature(payloadStr, secret);
      const result = await validateWebhookSignature(payloadStr, header, secret);

      expect(result).toEqual({
        valid: true,
        timestamp,
      });
    });

    it('should validate with Buffer payload', async () => {
      const { header, timestamp } = await generateSignature(payloadStr, secret);
      const payloadBuffer = Buffer.from(payloadStr, 'utf8');
      const result = await validateWebhookSignature(
        payloadBuffer,
        header,
        secret
      );

      expect(result).toEqual({
        valid: true,
        timestamp,
      });
    });

    it('should reject invalid signature', async () => {
      const { header } = await generateSignature(payloadStr, 'wrong_secret');
      const result = await validateWebhookSignature(payloadStr, header, secret);

      expect(result.valid).toBe(false);
      expect(result.error).toBe('Invalid signature');
    });

    it('should reject old timestamp (replay protection)', async () => {
      const oldTimestamp = Math.floor(Date.now() / 1000) - 400; // 400 seconds ago
      const { header } = await generateSignature(
        payloadStr,
        secret,
        oldTimestamp
      );
      const result = await validateWebhookSignature(payloadStr, header, secret);

      expect(result.valid).toBe(false);
      expect(result.error).toMatch(/Webhook timestamp too old/);
      expect(result.timestamp).toBe(oldTimestamp);
    });

    it('should reject future timestamp', async () => {
      const futureTimestamp = Math.floor(Date.now() / 1000) + 120; // 2 minutes in future
      const { header } = await generateSignature(
        payloadStr,
        secret,
        futureTimestamp
      );
      const result = await validateWebhookSignature(payloadStr, header, secret);

      expect(result.valid).toBe(false);
      expect(result.error).toBe('Webhook timestamp is in the future');
      expect(result.timestamp).toBe(futureTimestamp);
    });

    it('should allow disabling timestamp validation', async () => {
      const oldTimestamp = Math.floor(Date.now() / 1000) - 3600; // 1 hour ago
      const { header } = await generateSignature(
        payloadStr,
        secret,
        oldTimestamp
      );
      const result = await validateWebhookSignature(
        payloadStr,
        header,
        secret,
        {
          maxAgeSeconds: 0,
        }
      );

      expect(result).toEqual({
        valid: true,
        timestamp: oldTimestamp,
      });
    });

    it('should use custom max age', async () => {
      const timestamp = Math.floor(Date.now() / 1000) - 60; // 60 seconds ago
      const { header } = await generateSignature(payloadStr, secret, timestamp);

      // Should fail with 30 second max age
      const result1 = await validateWebhookSignature(
        payloadStr,
        header,
        secret,
        {
          maxAgeSeconds: 30,
        }
      );
      expect(result1.valid).toBe(false);

      // Should succeed with 120 second max age
      const result2 = await validateWebhookSignature(
        payloadStr,
        header,
        secret,
        {
          maxAgeSeconds: 120,
        }
      );
      expect(result2.valid).toBe(true);
    });

    it('should reject malformed signature header', async () => {
      const result = await validateWebhookSignature(
        payloadStr,
        'invalid_header',
        secret
      );
      expect(result).toEqual({
        valid: false,
        error: 'Invalid signature header format',
      });
    });

    it('should reject non-numeric timestamp', async () => {
      const header = 't=not_a_number,sha256=abcdef123456';
      const result = await validateWebhookSignature(payloadStr, header, secret);
      expect(result).toEqual({
        valid: false,
        error: 'Invalid timestamp in signature',
      });
    });

    it('should handle different payload modifications', async () => {
      const { header } = await generateSignature(payloadStr, secret);

      // Modified payload should fail
      const modifiedPayload = payloadStr.replace('main', 'master');
      const result1 = await validateWebhookSignature(
        modifiedPayload,
        header,
        secret
      );
      expect(result1.valid).toBe(false);

      // Extra whitespace should fail
      const result2 = await validateWebhookSignature(
        payloadStr + ' ',
        header,
        secret
      );
      expect(result2.valid).toBe(false);

      // Different encoding should work if content is same
      const payloadBuffer = Buffer.from(payloadStr);
      const result3 = await validateWebhookSignature(
        payloadBuffer,
        header,
        secret
      );
      expect(result3.valid).toBe(true);
    });
  });

  describe('validateWebhook', () => {
    it('should validate and parse webhook', async () => {
      const { header, timestamp } = await generateSignature(payloadStr, secret);
      const headers = {
        'x-pierre-signature': header,
        'x-pierre-event': 'push',
      };

      const result = await validateWebhook(payloadStr, headers, secret);

      expect(result.valid).toBe(true);
      expect(result.eventType).toBe('push');
      expect(result.timestamp).toBe(timestamp);
      expect(result.payload).toEqual(expectedPushPayload);
    });

    it('should handle uppercase headers', async () => {
      const { header, timestamp } = await generateSignature(payloadStr, secret);
      const headers = {
        'X-Pierre-Signature': header,
        'X-Pierre-Event': 'push',
      };

      const result = await validateWebhook(payloadStr, headers, secret);

      expect(result.valid).toBe(true);
      expect(result.eventType).toBe('push');
      expect(result.timestamp).toBe(timestamp);
    });

    it('should reject missing signature header', async () => {
      const headers = {
        'x-pierre-event': 'push',
      };

      const result = await validateWebhook(payloadStr, headers, secret);

      expect(result).toEqual({
        valid: false,
        error: 'Missing or invalid X-Pierre-Signature header',
      });
    });

    it('should reject missing event header', async () => {
      const { header } = await generateSignature(payloadStr, secret);
      const headers = {
        'x-pierre-signature': header,
      };

      const result = await validateWebhook(payloadStr, headers, secret);

      expect(result).toEqual({
        valid: false,
        error: 'Missing or invalid X-Pierre-Event header',
      });
    });

    it('should reject array headers', async () => {
      const { header } = await generateSignature(payloadStr, secret);

      const headers1 = {
        'x-pierre-signature': [header, header],
        'x-pierre-event': 'push',
      };
      const result1 = await validateWebhook(payloadStr, headers1, secret);
      expect(result1.valid).toBe(false);

      const headers2 = {
        'x-pierre-signature': header,
        'x-pierre-event': ['push', 'push'],
      };
      const result2 = await validateWebhook(payloadStr, headers2, secret);
      expect(result2.valid).toBe(false);
    });

    it('should reject invalid JSON payload', async () => {
      const invalidJson = 'not valid json';
      const { header } = await generateSignature(invalidJson, secret);
      const headers = {
        'x-pierre-signature': header,
        'x-pierre-event': 'push',
      };

      const result = await validateWebhook(invalidJson, headers, secret);

      expect(result.valid).toBe(false);
      expect(result.error).toBe('Invalid JSON payload');
    });

    it('should propagate signature validation errors', async () => {
      const { header } = await generateSignature(payloadStr, 'wrong_secret');
      const headers = {
        'x-pierre-signature': header,
        'x-pierre-event': 'push',
      };

      const result = await validateWebhook(payloadStr, headers, secret);

      expect(result.valid).toBe(false);
      expect(result.error).toBe('Invalid signature');
    });

    it('should pass through validation options', async () => {
      const oldTimestamp = Math.floor(Date.now() / 1000) - 3600; // 1 hour ago
      const { header } = await generateSignature(
        payloadStr,
        secret,
        oldTimestamp
      );
      const headers = {
        'x-pierre-signature': header,
        'x-pierre-event': 'push',
      };

      // Should fail with default max age
      const result1 = await validateWebhook(payloadStr, headers, secret);
      expect(result1.valid).toBe(false);

      // Should succeed with disabled timestamp validation
      const result2 = await validateWebhook(payloadStr, headers, secret, {
        maxAgeSeconds: 0,
      });
      expect(result2.valid).toBe(true);
      expect(result2.payload).toEqual(expectedPushPayload);
    });
  });

  describe('Security considerations', () => {
    it('should use constant-time comparison', async () => {
      // This test verifies the implementation uses timingSafeEqual
      // by ensuring different length signatures are rejected before comparison
      const { header } = await generateSignature(payloadStr, secret);
      const shortSigHeader = header.replace(/sha256=.*/, 'sha256=short');
      const result = await validateWebhookSignature(
        payloadStr,
        shortSigHeader,
        secret
      );

      expect(result.valid).toBe(false);
      expect(result.error).toBe('Invalid signature');
    });

    it('should handle empty or undefined inputs safely', async () => {
      const { header } = await generateSignature(payloadStr, secret);

      // Empty payload
      expect((await validateWebhookSignature('', header, secret)).valid).toBe(
        false
      );

      // Empty secret
      expect(
        (await validateWebhookSignature(payloadStr, header, '')).valid
      ).toBe(false);

      // Empty header
      expect(
        (await validateWebhookSignature(payloadStr, '', secret)).valid
      ).toBe(false);
    });

    it('should be resilient to timing attacks', async () => {
      // Generate multiple signatures to test timing consistency
      const signatures: string[] = [];
      for (let i = 0; i < 10; i++) {
        const testSecret = `secret_${i}`;
        const { header } = await generateSignature(payloadStr, testSecret);
        signatures.push(header);
      }

      // All invalid signatures should be rejected
      // The implementation should use constant-time comparison
      for (const sig of signatures) {
        const result = await validateWebhookSignature(payloadStr, sig, secret);
        expect(result.valid).toBe(false);
      }
    });
  });
});

File: packages/git-storage-sdk-go/webhook_test.go
package storage

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"net/http"
	"strconv"
	"strings"
	"testing"
	"time"
)

func TestParseSignatureHeader(t *testing.T) {
	header := "t=1234567890,sha256=abcdef123456"
	result := ParseSignatureHeader(header)
	if result == nil || result.Timestamp != "1234567890" || result.Signature != "abcdef123456" {
		t.Fatalf("unexpected signature header parse")
	}

	result = ParseSignatureHeader("t=1234567890, sha256=abcdef123456")
	if result == nil || result.Signature != "abcdef123456" {
		t.Fatalf("expected signature with spaces")
	}

	if ParseSignatureHeader("") != nil {
		t.Fatalf("expected nil for empty header")
	}
	if ParseSignatureHeader("invalid") != nil {
		t.Fatalf("expected nil for invalid header")
	}
	if ParseSignatureHeader("t=123") != nil {
		t.Fatalf("expected nil for missing signature")
	}
	if ParseSignatureHeader("sha256=abc") != nil {
		t.Fatalf("expected nil for missing timestamp")
	}
	if ParseSignatureHeader("timestamp=123,signature=abc") != nil {
		t.Fatalf("expected nil for wrong keys")
	}

	header = "t=1234567890,sha256=abcdef123456,v1=ignored"
	result = ParseSignatureHeader(header)
	if result == nil || result.Signature != "abcdef123456" {
		t.Fatalf("expected signature with extra fields")
	}
}

func TestValidateWebhookSignature(t *testing.T) {
	secret := "test_webhook_secret_key_123"
	payload := []byte(`{"repository":{"id":"repo","url":"https://git.example.com/org/repo"},"ref":"main","before":"abc","after":"def","customer_id":"cust","pushed_at":"2024-01-20T10:30:00Z"}`)
	stamp := time.Now().Unix()
	header := buildSignatureHeader(t, payload, secret, stamp)

	result := ValidateWebhookSignature(payload, header, secret, WebhookValidationOptions{})
	if !result.Valid || result.Timestamp != stamp {
		t.Fatalf("expected valid signature")
	}

	invalidHeader := buildSignatureHeader(t, payload, "wrong_secret", stamp)
	result = ValidateWebhookSignature(payload, invalidHeader, secret, WebhookValidationOptions{})
	if result.Valid || result.Error != "invalid signature" {
		t.Fatalf("expected invalid signature")
	}

	oldStamp := time.Now().Add(-400 * time.Second).Unix()
	header = buildSignatureHeader(t, payload, secret, oldStamp)
	result = ValidateWebhookSignature(payload, header, secret, WebhookValidationOptions{})
	if result.Valid || !strings.Contains(result.Error, "webhook timestamp too old") {
		t.Fatalf("expected old timestamp error")
	}

	futureStamp := time.Now().Add(120 * time.Second).Unix()
	header = buildSignatureHeader(t, payload, secret, futureStamp)
	result = ValidateWebhookSignature(payload, header, secret, WebhookValidationOptions{})
	if result.Valid || result.Error != "webhook timestamp is in the future" {
		t.Fatalf("expected future timestamp error")
	}

	stamp = time.Now().Add(-60 * time.Second).Unix()
	header = buildSignatureHeader(t, payload, secret, stamp)
	result = ValidateWebhookSignature(payload, header, secret, WebhookValidationOptions{MaxAgeSeconds: 30})
	if result.Valid {
		t.Fatalf("expected signature to be too old")
	}
	result = ValidateWebhookSignature(payload, header, secret, WebhookValidationOptions{MaxAgeSeconds: 120})
	if !result.Valid {
		t.Fatalf("expected signature to be valid with relaxed max age")
	}

	result = ValidateWebhookSignature(payload, "invalid_header", secret, WebhookValidationOptions{})
	if result.Valid || result.Error != "invalid signature header format" {
		t.Fatalf("expected invalid header format error")
	}

	result = ValidateWebhookSignature(payload, "t=not_a_number,sha256=abcdef", secret, WebhookValidationOptions{})
	if result.Valid || result.Error != "invalid timestamp in signature" {
		t.Fatalf("expected invalid timestamp error")
	}

	modified := []byte(strings.ReplaceAll(string(payload), "main", "master"))
	result = ValidateWebhookSignature(modified, header, secret, WebhookValidationOptions{})
	if result.Valid {
		t.Fatalf("expected modified payload to fail")
	}

	result = ValidateWebhookSignature(append(payload, ' '), header, secret, WebhookValidationOptions{})
	if result.Valid {
		t.Fatalf("expected whitespace payload to fail")
	}
}

func TestValidateWebhook(t *testing.T) {
	secret := "test_webhook_secret_key_123"
	payload := []byte(`{"repository":{"id":"repo_abc123","url":"https://git.example.com/org/repo"},"ref":"main","before":"abc123","after":"def456","customer_id":"cust_123","pushed_at":"2024-01-20T10:30:00Z"}`)
	stamp := time.Now().Unix()
	header := buildSignatureHeader(t, payload, secret, stamp)

	headers := http.Header{}
	headers.Set("x-pierre-signature", header)
	headers.Set("x-pierre-event", "push")

	result := ValidateWebhook(payload, headers, secret, WebhookValidationOptions{})
	if !result.Valid || result.EventType != "push" {
		t.Fatalf("expected valid webhook")
	}
	if result.Payload == nil || result.Payload.Push == nil {
		t.Fatalf("expected push payload")
	}
	if result.Payload.Push.CustomerID != "cust_123" || result.Payload.Push.Repository.ID != "repo_abc123" {
		t.Fatalf("unexpected push payload contents")
	}

	uppercase := http.Header{}
	uppercase.Set("X-Pierre-Signature", header)
	uppercase.Set("X-Pierre-Event", "push")
	result = ValidateWebhook(payload, uppercase, secret, WebhookValidationOptions{})
	if !result.Valid {
		t.Fatalf("expected valid webhook with uppercase headers")
	}

	missingSig := http.Header{}
	missingSig.Set("x-pierre-event", "push")
	result = ValidateWebhook(payload, missingSig, secret, WebhookValidationOptions{})
	if result.Valid || result.Error != "missing or invalid X-Pierre-Signature header" {
		t.Fatalf("expected missing signature error")
	}

	missingEvent := http.Header{}
	missingEvent.Set("x-pierre-signature", header)
	result = ValidateWebhook(payload, missingEvent, secret, WebhookValidationOptions{})
	if result.Valid || result.Error != "missing or invalid X-Pierre-Event header" {
		t.Fatalf("expected missing event error")
	}

	invalidJSON := []byte("not valid json")
	badHeader := buildSignatureHeader(t, invalidJSON, secret, stamp)
	badHeaders := http.Header{}
	badHeaders.Set("x-pierre-signature", badHeader)
	badHeaders.Set("x-pierre-event", "push")
	result = ValidateWebhook(invalidJSON, badHeaders, secret, WebhookValidationOptions{})
	if result.Valid || result.Error != "invalid JSON payload" {
		t.Fatalf("expected invalid JSON payload error")
	}

	wrongSig := buildSignatureHeader(t, payload, "wrong_secret", stamp)
	wrongHeaders := http.Header{}
	wrongHeaders.Set("x-pierre-signature", wrongSig)
	wrongHeaders.Set("x-pierre-event", "push")
	result = ValidateWebhook(payload, wrongHeaders, secret, WebhookValidationOptions{})
	if result.Valid || result.Error != "invalid signature" {
		t.Fatalf("expected invalid signature error")
	}
}

func TestWebhookEmptyInputs(t *testing.T) {
	secret := "test_webhook_secret_key_123"
	payload := []byte(`{"repository":{"id":"repo","url":"https://git.example.com/org/repo"},"ref":"main","before":"abc","after":"def","customer_id":"cust","pushed_at":"2024-01-20T10:30:00Z"}`)
	stamp := time.Now().Unix()
	header := buildSignatureHeader(t, payload, secret, stamp)

	result := ValidateWebhookSignature([]byte{}, header, secret, WebhookValidationOptions{})
	if result.Valid {
		t.Fatalf("expected empty payload to fail")
	}

	result = ValidateWebhookSignature(payload, header, "", WebhookValidationOptions{})
	if result.Valid || result.Error != "empty secret is not allowed" {
		t.Fatalf("expected empty secret error")
	}

	result = ValidateWebhookSignature(payload, "", secret, WebhookValidationOptions{})
	if result.Valid || result.Error != "invalid signature header format" {
		t.Fatalf("expected empty header error")
	}
}

func buildSignatureHeader(t *testing.T, payload []byte, secret string, timestamp int64) string {
	t.Helper()
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(strconv.FormatInt(timestamp, 10) + "." + string(payload)))
	signature := hex.EncodeToString(mac.Sum(nil))
	return "t=" + strconv.FormatInt(timestamp, 10) + ",sha256=" + signature
}
